### PR TITLE
Issue/481 crud task templates

### LINKED ISSUE #479: Load ward task templates

### LINKED ISSUE #480: Load personal task templates

### LINKED ISSUE #481: Create, update and delete task templates

### PR DESCRIPTION
**Pull request by florianjs**
## Which issues does this pull request close?
closes #479
closes #480 
closes #481

### CODE REVIEW

File: tasks/components/SubtaskTile.tsx [hunk 1]
```diff
@@ -47,7 +47,7 @@ export const SubtaskTile = ({
       <div>
         <Checkbox
           onChange={value => onChange({ ...subtask, isDone: value })}
-          checked={subtask.isDone}
+          checked={subtask.isDone !== undefined ? subtask.isDone : false}
```
Comment (use-to) [inline tasks/components/SubtaskTile.tsx:50]:
```suggestion
          checked={!!subtask.isDone}
```

**Review by use-to — CHANGES_REQUESTED** [review]

File: tasks/components/SubtaskTile.tsx (hunk 1 above)
Comment (use-to) [reply tasks/components/SubtaskTile.tsx:50]: @florianjs already marked it as resolved?

Comment (florianjs) [reply tasks/components/SubtaskTile.tsx]: Yes, it just hasn't been on github yet.

**Review by use-to — CHANGES_REQUESTED** [review]
As we updated our node version and different packages affecting our linting and ci, please merge main in to this branch.

**Comment by use-to** [comment]
@MaxSchaefer Similar to the task creation, notes are required for the request to be successful.

**Comment by use-to** [comment]
@florianjs This looks really promising. Great work, I tried to reproduce a bug regarding the `queryKey`'s... Will test it later on again.

**Comment by DasProffi** [comment]
> @MaxSchaefer Similar to the task creation, notes are required for the request to be successful.

I created an issue for this https://github.com/helpwave/services/issues/289

**Comment by MaxSchaefer** [comment]
@florianjs and I migrated this to be used with context. @florianjs is working on the last to-do: creating additional subtasks during an update of a task template. After that, this pull request is ready for review.

**Review by MaxSchaefer — CHANGES_REQUESTED** [review]
The delete behavior of a subtasks differs between a ward task template and a personal one. A deletion of a subtask in a personal task template is reflected directly. A deletion of a subtask in a ward task template not, here I need to press "update" first. I prefer the later one.

**Review by use-to — CHANGES_REQUESTED** [review]
[buggy_view_task_templates.webm](https://github.com/helpwave/web/assets/13922835/62c609f5-a327-41f0-b79b-93335eca6ffe)

Buggy view for ward task templates.
Used: http://localhost:3000/ward/f0509ffa-801f-4e85-91eb-c7b90dd65ff6/templates

Personal Task Templates works flawlessly for me! 👍🏻

**Comment by florianjs** [comment]
> First of all: Personal Task Templates works flawlessly for me! 👍🏻
> 
>  buggy_view_task_templates.webm 
> Buggy view for ward task templates. Used: http://localhost:3000/ward/f0509ffa-801f-4e85-91eb-c7b90dd65ff6/templates
> 
> 1. Can't create ward task template
> 2. Subtasks can be checked but view will not be updated!

Sorry, I didn't pay attention to it, but it should be fixed.

File: package-lock.json
Comment (use-to) [inline package-lock.json:1]: Delete this file `package-lock.json`, please!
We are using pnpm. This is not part of our project setup.

File: tasks/mutations/task_mutations.ts [hunk 2]
```diff
@@ -13,10 +12,10 @@ import {
   GetTaskRequest,
   UpdateTaskRequest,
   SubTaskToDoneRequest,
-  SubTaskToToDoRequest
+  SubTaskToToDoRequest, GetTasksByPatientSortedByStatusRequest
```
Comment (use-to) [inline tasks/mutations/task_mutations.ts:15]:
```suggestion
  SubTaskToToDoRequest, 
  GetTasksByPatientSortedByStatusRequest,
```

File: tasks/components/SubtaskView.tsx [hunk 3]
```diff
@@ -51,19 +52,17 @@ export const SubtaskView = ({
...
 
-  const isCreatingTask = !!taskID
-  const addSubtaskMutation = useSubTaskAddMutation(() => undefined, taskID)
-  const updateSubtaskMutation = useSubTaskUpdateMutation()
-  const deleteSubtaskMutation = useSubTaskDeleteMutation()
-  const subtaskToToDoMutation = useSubTaskToToDoMutation()
-  const subtaskToDoneMutation = useSubTaskToDoneMutation()
-
+  // TODO: Remove ?? '' once the mutate functions are passed properly
+  useSubTaskTemplateAddMutation(() => undefined, taskTemplateId ?? '')
```
Comment (use-to) [inline tasks/components/SubtaskView.tsx:65]: As you are using `noop` as default parameter, just leave it out.
```suggestion
  useSubTaskTemplateAddMutation(taskTemplateId ?? '')
```

**Review by use-to — CHANGES_REQUESTED** [review]
Very good - the waiting has been worthwhile, everything works like a charm! 🥳 

Please look over the following small requests. I am convinced that we can merge this PR today.

**Comment by use-to** [comment]
@MaxSchaefer From my side this pull request is done. As you requested changes, too: Can you take a look real quick?

**Review by use-to — APPROVED** [review]
LGTM!

**Comment by use-to** [comment]
> First of all: Personal Task Templates works flawlessly for me! 👍🏻
> 
>  buggy_view_task_templates.webm 
> Buggy view for ward task templates. Used: http://localhost:3000/ward/f0509ffa-801f-4e85-91eb-c7b90dd65ff6/templates
> 
> 1. Can't create ward task template
> 2. Subtasks can be checked but view will not be updated!

Will be fixed in #519 - not part of this pr anymore as it is detail work which does not directly affect the functionality of this application

**Review by MaxSchaefer — DISMISSED** [review]
Works and looks great! 🚀🌊
Thank you very much @florianjs!

**Review by use-to — APPROVED** [review]

**Pull request merged**